docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+            openPublicPortal(<publicMagicIp>, port.port, port.hostPort)
+          }
+        } else if service.spec.visibility == "" {
+          for each service.publicIP {  // deprecated
+            openPublicPortal(publicIP, port.port, proxyPort) // uses cloud-provider to decide srcip vs dstip (gclb vs rax)
+          }
+        }
+   }
+}
+```
+
+
+## kube-proxy iptables
+
+As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by
+iptables.  These ports are not visible other than to kube-proxy & iptables.  We need to be sure that these ports will
+not conflict with the cluster-wide ports.  To do that, we must make sure that kube-proxy differentiates between
+internal & external traffic, and uses these internal ports only with internal traffic.
+
+There are a few possible implementation strategies here:
+
+1. Use `sysctl -w net.ipv4.conf.all.route_localnet=1`, and then bind kube-proxy to 127.0.0.1 for internal traffic.  Requires kernel >= 3.6
+1. Use addrtype with iptables (external: `! --src-type local --dst-type local`, internal: `--src-type local --dst-type local`)
+1. Use the network interface to identify traffic (internal: lo, cbr0; external: eth0)
+1. Use the container CIDR range to identify internal (container) traffic (e.g. 10.244.0.0/16).  Requires fixing that CIDR.
+
+I would like to try those strategies, basically working from top to bottom until we find that one that works.  I tried
+the container CIDR strategy, so that should work.
+
+The sysctl strategy is nice, but does not work on old kernels, and so I propose for those creating a dummy internal address.
+I think the easiest way to do that is to create a bridge and assign it an IP.  This is a hack, but should allow us
+to support old kernels until they are replaced.  We should also be able to put this hack in the Salt scripts.
+
+The sysctl strategy (at least) gives simpler iptables rule than what we have now, so should make a pure-iptables implementation
+of kube-proxy easier.
+
+It is possible that there would be an unexpected port-conflict meaning we could not open the port.  This will be treated
+as an unexpected failure (like out-of-fds); it will be logged and kube-proxy will continue and likely retry.  This should
+only happen if we run new services on the minions or choose a public_service_ports poorly.
+
+
+## Sidecar-pod implementations
+
+In future, we may allow sidecar-pods to be responsible for configuring load-balancers; these could be outside
+of the main k8s codebase.  We would likely configure the apiserver to permit load-balancer creation, assign a port
+(probably), but k8s would not create the load balancer.  kube-proxy would create the service as a public service,
+(possibly with additional firewall rules).  The loadbalancer sidecar-pod would create the load-balancer and
+call the API to report events.
+
+We could move the cloud-provider load-balancer functionality into a sidecar-pod, but not for V1.
```

Comment:
I'm not sure we can, as long as we need to feed kube-proxy with info on how to receive from those load-balancers.